test/e2e/shared/common.go

```diff
@@ -262,3 +265,125 @@ func SetEnvVar(key, value string, private bool) {
 	Logf("Setting environment variable: key=%s, value=%s", key, printableValue)
 	_ = os.Setenv(key, value)
 }
+
+// getOpenStackClusterFromMachine gets the OpenStackCluster that is related to the given machine.
+func getOpenStackClusterFromMachine(ctx context.Context, client client.Client, machine *clusterv1.Machine) (*infrav1.OpenStackCluster, error) {
+	key := types.NamespacedName{
+		Namespace: machine.Namespace,
+		Name:      machine.Spec.ClusterName,
+	}
+	cluster := &clusterv1.Cluster{}
+	err := client.Get(ctx, key, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	key = types.NamespacedName{
+		Namespace: cluster.Spec.InfrastructureRef.Namespace,
+		Name:      cluster.Spec.InfrastructureRef.Name,
+	}
+	openStackCluster := &infrav1.OpenStackCluster{}
+	err = client.Get(ctx, key, openStackCluster)
+	return openStackCluster, err
+}
+
+type OpenStackLogCollector struct {
+	E2EContext E2EContext
+}
+
+// CollectMachineLog gets logs for the OpenStack resources related to the given machine.
+func (o OpenStackLogCollector) CollectMachineLog(ctx context.Context, managementClusterClient client.Client, m *clusterv1.Machine, outputPath string) error {
+	machineLogBase := path.Join(outputPath, "instances", m.Namespace, m.Name)
+	metaLog := path.Join(machineLogBase, "instance.log")
+
+	if err := os.MkdirAll(filepath.Dir(metaLog), 0o750); err != nil {
+		_, _ = fmt.Fprintf(GinkgoWriter, "couldn't create directory %q for file: %s\n", metaLog, err)
+	}
+
+	f, err := os.OpenFile(metaLog, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
+	if err != nil {
+		_, _ = fmt.Fprintf(GinkgoWriter, "couldn't open file %q: %s\n", metaLog, err)
+		return nil
+	}
+	defer f.Close()
+
+	openStackCluster, err := getOpenStackClusterFromMachine(ctx, managementClusterClient, m)
+	if err != nil {
+		return fmt.Errorf("error getting OpenStackCluster for Machine: %s", err)
+	}
+
+	if (m.Status.Addresses == nil) || (len(m.Status.Addresses) < 1) {
```

Comment:
Go nit: idiomatically this would be
```suggestion
	if len(m.Status.Addresses) < 1 {
```
because the length of a nil slice is well defined (as zero). Incidentally it's also safe and idiomatic to iterate over a nil slice in a for loop.

Reply:
Thanks! Fixing